ttg/ttg/terminal.h

```diff
@@ -11,8 +11,80 @@
 #include "ttg/util/trace.h"
 #include "ttg/util/demangle.h"
 #include "ttg/world.h"
+#include "boost/callable_traits.hpp"
 
 namespace ttg {
+  namespace detail {
+
+    /* Wraps any key,value data structure.
+     * Elements of the data structure can be accessed using get method, which calls the at method of the Container.
+     * keyT - taskID
+     * valueT - Value type of the Container
+    */
+    template<typename keyT, typename valueT>
+    struct ContainerWrapper {
+      std::function<valueT (keyT const& key)> get = nullptr;
+      std::function<size_t (keyT const& key)> owner = nullptr;
+
+      ContainerWrapper() = default;
+
+      template<typename T, typename mapperT, typename keymapT, std::enable_if_t<!std::is_same<std::decay_t<T>,
+                                                          ContainerWrapper>{}, bool> = true>
+        //Store a pointer to the user's container in std::any, no copies
+        ContainerWrapper(T &t, mapperT &&mapper,
+                         keymapT &&keymap) : get([&t,&mapper](keyT const &key) {
+                                                  if constexpr (!std::is_class_v<T> && std::is_function_v<T>) {
+                                                      auto k = mapper(key);
+                                                      return t(k); //Call the user-defined lambda function.
+                                                    }
+                                                  else
+                                                    {
+                                                      auto k = mapper(key);
+                                                      using key_type = typename T::key_type;
+                                                      //at method returns a const ref to the item.
+                                                      return t.at(k);
+                                                    }
+                                                }),
+                                            owner([&t, &mapper, &keymap](keyT const &key) {
+                                                    auto idx = mapper(key); //Mapper to map task ID to index of the data structure.
+                                                    return keymap(idx);
+                                                  })
+        {
+         //Validate mapper and keymap function arguments
+         using mapper_args_t = boost::callable_traits::args_t<mapperT>;
+         constexpr auto num_args = std::tuple_size<mapper_args_t>::value;
+         static_assert(num_args == 1);
+         static_assert(std::is_same_v<typename std::tuple_element<0, mapper_args_t>::type, const keyT &>,
+                       "argument of Mapper must be const keyT& (unless keyT = void)");
+         using km_args_t = boost::callable_traits::args_t<keymapT>;
```

Comment:
That won't work with generic lambdas. Maybe `std::is_invocable_r` will work as well?

Reply:
Actually, we probably don't need to validate the functions here, the compiler will balk if the types don't pan out...